File: packages/expo-updates/ios/EXUpdates/SelectionPolicy/SelectionPolicyFactory.swift
//  Copyright © 2019 650 Industries. All rights reserved.

import Foundation

/**
 * Factory class to ease the construction of [SelectionPolicy] objects whose three methods all use
 * the same ordering policy.
 */
@objc(EXUpdatesSelectionPolicyFactory)
@objcMembers
public final class SelectionPolicyFactory: NSObject {
  public static func filterAwarePolicy(withRuntimeVersion runtimeVersion: String, config: UpdatesConfig, filterByChannel: Bool = false) -> SelectionPolicy {
    return SelectionPolicy.init(
      launcherSelectionPolicy: LauncherSelectionPolicyFilterAware.init(runtimeVersion: runtimeVersion, filterByChannel: filterByChannel, config: config),
      loaderSelectionPolicy: LoaderSelectionPolicyFilterAware(disableAntiBrickingMeasures: config.disableAntiBrickingMeasures),
      reaperSelectionPolicy: ReaperSelectionPolicyFilterAware()
    )
  }
}

File: packages/expo-updates/android/src/main/java/expo/modules/updates/selectionpolicy/SelectionPolicyFactory.kt
package expo.modules.updates.selectionpolicy

import expo.modules.updates.UpdatesConfiguration

/**
 * Factory class to ease the construction of [SelectionPolicy] objects whose three methods all use
 * the same ordering policy.
 */
object SelectionPolicyFactory {
  @JvmStatic fun createFilterAwarePolicy(
    runtimeVersion: String, 
    config: UpdatesConfiguration? = null,
    filterByChannel: Boolean = false
  ): SelectionPolicy {
    // When disableAntiBrickingMeasures is enabled, pass it to the loader selection policy
    // to allow loading updates with older commit times (needed for channel switching)
    val disableAntiBrickingMeasures = config?.disableAntiBrickingMeasures ?: false
    return SelectionPolicy(
      LauncherSelectionPolicyFilterAware(runtimeVersion, config, filterByChannel),
      LoaderSelectionPolicyFilterAware(disableAntiBrickingMeasures),
      ReaperSelectionPolicyFilterAware()
    )
  }
}
